tools/deno_http_proxy.ts

```diff
@@ -21,6 +21,10 @@ async function proxyRequest(req: ServerRequest) {
     method: req.method,
     headers: req.headers
   });
+  // TODO(kt3k): lib.deno_runtime.d.ts has 2 EOF types: Deno.EOF and io.EOF.
+  // They are identical symbols, and should be compatible. However typescript
+  // recognizes they are different types and the below call doesn't compile.
+  // @ts-ignore
   req.respond(resp);
```

Comment:
That's annoying.

@kitsonk can you advise us?

Reply:
@kitsonk 
I have trouble with this line. The error is like the below:
<img width="1434" alt="スクリーンショット 2019-07-02 0 01 35" src="https://user-images.githubusercontent.com/613956/60446556-96a79480-9c5c-11e9-9fef-54d01e91a1f9.png">

[lib.deno_runtime.d.ts](https://gist.github.com/kt3k/8cd3f563dc2527cb1552f0d3d315ba8f) of this branch contains two EOF `unique symbol` declarations. (This seems strange because in this branch EOF is defined only once, but .d.ts treat it as different types depending on where it is used.)

I tried to move EOF to several other places (js/eof.ts, js/deno.ts, etc) to avoid this, but I couldn't find  a solution.

Reply:
@kt3k I just checked out this branch locally and it compiles `tools/deno_http_proxy.ts` without any problem.

Reply:
@kt3k If you define two identical symbols there are still not equal:
```js
const foo: unique symbol = Symbol("foo");
const foo1: unique symbol = Symbol("foo");

console.log(foo === foo1);

// output
error TS2367: This condition will always return 'false' since the types 'typeof foo' and 'typeof foo1' have no overlap.

► file:///Users/biwanczuk/dev/deno/test.ts:4:13

4 console.log(foo === foo1);
```

That's what symbols are for - to get unique value if another symbol with this name is created.
Correct me if I'm wrong but I believe you can't solve this issue without ignoring type check temporarily.

Reply:
Probably doing something in deno_std like
```ts
export const EOF: unique symbol = Deno.EOF || Symbol("EOF");
export type EOF = typeof EOF;
```
?
(I did not follow the context of this discussion so I might be wrong)

Reply:
The problem seems more subtle to me.
This change doesn't include 2 symbol definitions. EOF is defined only in `io.ts`. `deno.ts` import it and re-export it. In that case, EOF in io.ts and EOF in deno.ts should be the same thing. But lib.deno_runtime.d.ts (= `deno types`) includes 2 EOFs under `namespace Deno` and `namespace io`. Actually lib.deno_runtime.d.ts includes the entire contents of `io.ts` twice in Deno and io namespaces. (io namespace is not accessible from the user. I'm not sure why this exists. Probably a workaround for something.)

lib.deno_runtime.d.ts is created by the special tool `//tools/ts_library_builder/`. I suspect something goes wrong in it.

Reply:
Found a workaround for this:

```ts
// TODO: Make this symbol type
export const EOF: "EOF" = "EOF";
export type EOF = typeof EOF;
```

This avoids the above problem.

Reply:
Does it even need to be a symbol though?

Reply:
@j-f1 
I think symbol type is ideal for the type of EOF and we need to make it symbol in the future. But for now I don't see any way to declare it as symbol correctly, so I would suggest the above as an alternative path to make things forward.

Reply:
@kt3k Does that mean you're ready to land https://github.com/denoland/deno/pull/2591 ?

Reply:
Found another workaround. If we move `EOF` to global namespace (window), it seemed working. The diff looks like this ( https://github.com/kt3k/deno/commit/d08465dcaac477060bcecfee0e0fd56dd14fef9e )

Reply:
@kt3k TBH I don't understand the issue, but I'm fine with that workaround (it's better than making it global). I think we should just move forward with this change and try to clean it up in later. I suspect some changes to ts_library_builder could fix this...

Reply:
Could you do `export const EOF: typeof Deno.EOF = Deno.EOF` in `io.ts`?

Reply:
@ry OK. I'll update the PR to use the workaround.

> I suspect some changes to ts_library_builder could fix this.

I agree that this needs some changes to ts_library_builder.

@j-f1 
That causes the following error:
```
python ../../tools/run_node.py ../../node_modules/ts-node/dist/bin.js --project /Users/kt3k/s/deno/tools/ts_library_builder/tsconfig.json --skip-ignore ../../tools/ts_library_builder/main.ts --basePath ../../ --inline ../../js/lib.web_assembly.d.ts --buildPath . --outFile gen/cli/lib/lib.deno_runtime.d.ts --silent --debug
../../js/io.ts:6:26 - error TS4025: Exported variable 'EOF' has or is using private name 'Deno'.

6 export const EOF: typeof Deno.EOF = Deno.EOF;
                           ~~~~
```